File: src/services/weatherService.ts

interface WeatherData {
  temperature: string;
  humidity: string;
  windSpeed: string;
  precipitation: string;
  location: string;
  description?: string;
  pressure?: string;
  visibility?: string;
  cloudCover?: string;
  feelsLike?: string;
  weatherIcon?: string;
}

class WeatherService {
  private apiKey = import.meta.env.VITE_OPENWEATHER_API_KEY;
  private baseUrl = 'https://api.openweathermap.org/data/2.5/weather';
  private lastRealDataTime: Date | null = null;

  async getRealTimeWeatherData(location: 'thies' | 'taiba-ndiaye' | 'hann-maristes' | 'dakar' | 'bargny'): Promise<WeatherData | null> {
    if (!this.apiKey) {
      console.warn('⚠️ Clé API OpenWeather manquante - utilisation données de secours uniquement');
      return this.getFallbackData(location);
    }

    const cityMapping = {
      'thies': 'Thiès,SN',
      'taiba-ndiaye': 'Taiba Ndiaye,SN', 
      'hann-maristes': 'Hann,SN',
      'dakar': 'Dakar,SN',
      'bargny': 'Bargny,SN'
    };

    const cityQuery = cityMapping[location] || 'Thiès,SN';
    
    try {
      const url = `${this.baseUrl}?q=${encodeURIComponent(cityQuery)}&appid=${this.apiKey}&units=metric&lang=fr`;
      console.log(`🌍 Récupération données météo temps réel pour ${location}...`);
      
      const response = await fetch(url, { 
        headers: {
          'Accept': 'application/json'
        }
      });

      if (!response.ok) {
        throw new Error(`API Error: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      
      // Valider que les données sont complètes
      if (!data.main || !data.weather || !data.wind) {
        throw new Error('Données API incomplètes');
      }

      this.lastRealDataTime = new Date();
      const weatherData = this.formatOpenWeatherData(data);
      
      console.log(`✅ Données météo temps réel récupérées pour ${location}:`, weatherData);
      return weatherData;

    } catch (error) {
      console.error(`❌ Échec récupération données temps réel pour ${location}:`, error);
      
      // Utiliser les données de secours uniquement en cas d'échec
      console.log(`🔄 Basculement vers données de secours pour ${location}`);
      return this.getFallbackData(location);
    }
  }

  private formatOpenWeatherData(data: any): WeatherData {
    // Calcul précipitations (pluie + neige sur 1h)
    let precipitation = 0;
    if (data.rain?.['1h']) precipitation += data.rain['1h'];
    if (data.snow?.['1h']) precipitation += data.snow['1h'];

    return {
      temperature: `${Math.round(data.main.temp)}°C`,
      humidity: `${data.main.humidity}%`,
      windSpeed: `${Math.round(data.wind.speed * 3.6)} km/h`,
      precipitation: `${precipitation.toFixed(1)} mm`,
      location: data.name,
      description: data.weather[0].description.charAt(0).toUpperCase() + data.weather[0].description.slice(1),
      pressure: `${data.main.pressure} hPa`,
      feelsLike: `${Math.round(data.main.feels_like)}°C`,
      visibility: data.visibility ? `${(data.visibility / 1000).toFixed(1)} km` : undefined,
      cloudCover: data.clouds ? `${data.clouds.all}%` : undefined,
      weatherIcon: this.getWeatherIcon(data.weather[0].icon)
    };
  }

  private getWeatherIcon(iconCode: string): string {
    const iconMapping = {
      "01d": "sun", "01n": "moon",
      "02d": "cloud", "02n": "cloud",
      "03d": "cloud", "03n": "cloud",
      "04d": "cloud", "04n": "cloud",
      "09d": "rain", "09n": "rain",
      "10d": "rain", "10n": "rain",
      "11d": "storm", "11n": "storm",
      "13d": "snow", "13n": "snow",
      "50d": "mist", "50n": "mist"
    };
    return iconMapping[iconCode] || "sun";
  }

  private getFallbackData(location: string): WeatherData {
    const locationNames = {
      'thies': 'Thiès',
      'taiba-ndiaye': 'Taïba Ndiaye',
      'hann-maristes': 'Hann Maristes', 
      'dakar': 'Dakar',
      'bargny': 'Bargny'
    };

    // Données de secours réalistes pour la région
    const currentMonth = new Date().getMonth() + 1;
    const isDrySeason = [11, 12, 1, 2, 3, 4, 5].includes(currentMonth);
    
    // Variation légère pour simuler des conditions changeantes
    const tempVariation = Math.floor(Math.random() * 4) - 2;
    const humidityVariation = Math.floor(Math.random() * 10) - 5;
    
    if (isDrySeason) {
      return {
        temperature: `${28 + tempVariation}°C`,
        humidity: `${55 + humidityVariation}%`,
        windSpeed: `${12 + Math.floor(Math.random() * 6)} km/h`,
        precipitation: `${(Math.random() * 1.2).toFixed(1)} mm`,
        location: locationNames[location as keyof typeof locationNames] || 'Thiès',
        description: 'Données locales - Saison sèche',
        pressure: `${1014 + Math.floor(Math.random() * 8)} hPa`,
        feelsLike: `${30 + tempVariation}°C`,
        visibility: `${10 + Math.floor(Math.random() * 5)} km`,
        cloudCover: `${Math.floor(Math.random() * 30)}%`,
        weatherIcon: 'sun'
      };
    } else {
      return {
        temperature: `${25 + tempVariation}°C`,
        humidity: `${78 + humidityVariation}%`,
        windSpeed: `${10 + Math.floor(Math.random() * 8)} km/h`,
        precipitation: `${(8 + Math.random() * 15).toFixed(1)} mm`,
        location: locationNames[location as keyof typeof locationNames] || 'Thiès',
        description: 'Données locales - Saison des pluies',
        pressure: `${1010 + Math.floor(Math.random() * 6)} hPa`,
        feelsLike: `${27 + tempVariation}°C`,
        visibility: `${6 + Math.floor(Math.random() * 4)} km`,
        cloudCover: `${60 + Math.floor(Math.random() * 30)}%`,
        weatherIcon: 'rain'
      };
    }
  }

  // Méthode pour données de secours (legacy)
  async getWeatherData(location: string): Promise<WeatherData> {
    return this.getFallbackData(location);
  }

  isUsingRealData(): boolean {
    if (!this.lastRealDataTime) return false;
    const fiveMinutesAgo = new Date(Date.now() - 5 * 60 * 1000);
    return this.lastRealDataTime > fiveMinutesAgo;
  }
}

export const weatherService = new WeatherService();
export type { WeatherData };

File: backend/services/weather_service.py

import requests
import os
from config.database import log_weather

class WeatherService:
    def __init__(self):
        self.api_key = os.getenv("OPENWEATHER_API_KEY")
        self.base_url = "http://api.openweathermap.org/data/2.5/weather"
    
    def get_weather_data(self, location):
        """Récupère les données météo pour une ville donnée"""
        if not self.api_key:
            return self._get_fallback_data(location)
        
        # Mapping des villes
        city_mapping = {
            "thies": "Thiès,SN",
            "taiba-ndiaye": "Taiba Ndiaye,SN"
        }
        
        city = city_mapping.get(location.lower(), location)
        
        try:
            url = f"{self.base_url}?q={city}&appid={self.api_key}&units=metric&lang=fr"
            response = requests.get(url, timeout=10)
            
            if response.status_code == 200:
                data = response.json()
                weather_data = self._format_weather_data(data, location)
                
                # Log dans la base de données
                log_weather(
                    location=weather_data['location'],
                    temperature=float(weather_data['temperature'].replace('°C', '')),
                    humidity=float(weather_data['humidity'].replace('%', '')),
                    wind_speed=float(weather_data['windSpeed'].split()[0]),
                    precipitation=float(weather_data['precipitation'].replace(' mm', ''))
                )
                
                return weather_data
            else:
                print(f"⚠️ API Météo error {response.status_code}: {response.text}")
                return self._get_fallback_data(location)
                
        except Exception as e:
            print(f"❌ Erreur météo: {e}")
            return self._get_fallback_data(location)
    
    def _format_weather_data(self, data, location):
        """Formate les données météo de l'API"""
        return {
            "temperature": f"{round(data['main']['temp'])}°C",
            "humidity": f"{data['main']['humidity']}%",
            "windSpeed": f"{round(data['wind']['speed'] * 3.6)} km/h",
            "precipitation": f"{data.get('rain', {}).get('1h', 0)} mm",
            "weatherIcon": self._get_weather_icon(data['weather'][0]['icon']),
            "location": data['name']
        }
    
    def _get_weather_icon(self, icon_code):
        """Convertit le code icône OpenWeather en emoji"""
        icon_mapping = {
            "01d": "sun", "01n": "moon",
            "02d": "cloud", "02n": "cloud",
            "03d": "cloud", "03n": "cloud",
            "04d": "cloud", "04n": "cloud",
            "09d": "rain", "09n": "rain",
            "10d": "rain", "10n": "rain",
            "11d": "storm", "11n": "storm",
            "13d": "snow", "13n": "snow",
            "50d": "mist", "50n": "mist"
        }
        return icon_mapping.get(icon_code, "sun")
    
    def _get_fallback_data(self, location):
        """Données de secours si l'API échoue"""
        location_names = {
            "thies": "Thiès",
            "taiba-ndiaye": "Taïba Ndiaye"
        }
        
        return {
            "temperature": "28°C",
            "humidity": "65%",
            "windSpeed": "12 km/h",
            "precipitation": "2.5 mm",
            "weatherIcon": "sun",
            "location": location_names.get(location, location.title())
        }

# Instance globale
weather_service = WeatherService()
